README.md
# Parity PMD

Parity PMD implements [Parity][]'s native market data protocol on Node.

  [Parity]: https://github.com/paritytrading/parity

Parity PMD requires Node 6 or newer and Parity 0.7.0 or newer.

## Install

Install Parity PMD:

```
npm install parity-pmd
```

## Links

For more information on Parity PMD:

- Follow [@paritytrading](https://twitter.com/paritytrading) on Twitter for
  news and announcements
- Join [paritytrading/chat](https://gitter.im/paritytrading/chat) on Gitter
  for discussions

## License

Copyright 2015 Jussi Virtanen.

Released under the Apache License, Version 2.0. See `LICENSE.txt` for details.


CHANGELOG.md
# Release Notes

## 0.4.0 (2017-08-11)

- Upgrade to Parity 0.7.0

## 0.3.0 (2017-04-14)

- Update repository

## 0.2.0 (2017-03-19)

- Remove `long` dependency
- Handle too short strings

## 0.1.1 (2015-06-30)

- Add description to package specification

## 0.1.0 (2015-06-29)

- Initial release


test.js
'use strict';

const PMD = require('./');
const assert = require('assert');

describe('PMD', function () {
  const messages = [
    {
      name: 'Version',
      formatted: [
        0x56,
        0x00, 0x00, 0x00, 0x01,
      ],
      parsed: {
        messageType: 'V',
        version: 1,
      },
    },
    {
      name: 'Order Added',
      formatted: [
        0x41,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
        0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x02,
        0x42,
        0x46, 0x4f, 0x4f, 0x20, 0x20, 0x20, 0x20, 0x20,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x03,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04,
      ],
      parsed: {
        messageType: 'A',
        timestamp: 1,
        orderNumber: 4294967298,
        side: 'B',
        instrument: 'FOO     ',
        quantity: 3,
        price: 4,
      },
    },
    {
      name: 'Order Executed',
      formatted: [
        0x45,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
        0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x02,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x03,
        0x00, 0x00, 0x00, 0x04,
      ],
      parsed: {
        messageType: 'E',
        timestamp: 1,
        orderNumber: 4294967298,
        quantity: 3,
        matchNumber: 4,
      },
    },
    {
      name: 'Order Canceled',
      formatted: [
        0x58,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
        0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x02,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x03,
      ],
      parsed: {
        messageType: 'X',
        timestamp: 1,
        orderNumber: 4294967298,
        canceledQuantity: 3,
      },
    },
  ];

  describe('#format()', function () {
    messages.forEach((message) => {
      it(`handles ${message.name} message`, function () {
        assert.deepEqual(PMD.format(message.parsed), Buffer.from(message.formatted));
      });
    });

    it('handles unknown message type', function () {
      const message = {
        messageType: '?'
      };

      assert.throws(() => PMD.format(message), /Unknown message type: \?/);
    });

    it('handles too short string', function () {
      const formatted = [
        0x41,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02,
        0x42,
        0x46, 0x4f, 0x4f, 0x20, 0x20, 0x20, 0x20, 0x20,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x03,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04,
      ];

      const parsed = {
        messageType: 'A',
        timestamp: 1,
        orderNumber: 2,
        side: 'B',
        instrument: 'FOO',
        quantity: 3,
        price: 4,
      };

      assert.deepEqual(PMD.format(parsed), Buffer.from(formatted));
    });

    it('handles too long string', function () {
      const formatted = [
        0x41,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02,
        0x42,
        0x46, 0x4f, 0x4f, 0x20, 0x42, 0x41, 0x52, 0x20,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x03,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04,
      ];

      const parsed = {
        messageType: 'A',
        timestamp: 1,
        orderNumber: 2,
        side: 'B',
        instrument: 'FOO BAR BAZ',
        quantity: 3,
        price: 4,
      }

      assert.deepEqual(PMD.format(parsed), Buffer.from(formatted));
    });
  });

  describe('#parse()', function () {
    messages.forEach((message) => {
      it(`handles ${message.name} message`, function () {
        assert.deepEqual(PMD.parse(Buffer.from(message.formatted)), message.parsed);
      });
    });

    it('handles unknown message type', function () {
      const buffer = Buffer.from([ 0x3f ]);

      assert.throws(() => PMD.parse(buffer), /Unknown message type: 63/);
    });
  });
});


index.js
'use strict';

exports.format = (message) => {
  switch (message.messageType) {
    case 'V':
      return formatVersion(message);
    case 'A':
      return formatOrderAdded(message);
    case 'E':
      return formatOrderExecuted(message);
    case 'X':
      return formatOrderCanceled(message);
    default:
      throw new Error('Unknown message type: ' + message.messageType);
  }
};

exports.parse = (buffer) => {
  const messageType = buffer.readUInt8(0);

  switch (messageType) {
    case 0x56:
      return parseVersion(buffer);
    case 0x41:
      return parseOrderAdded(buffer);
    case 0x45:
      return parseOrderExecuted(buffer);
    case 0x58:
      return parseOrderCanceled(buffer);
    default:
      throw new Error('Unknown message type: ' + messageType);
  }
};

function formatVersion(message) {
  const buffer = Buffer.allocUnsafe(5);

  buffer.writeUInt8(0x56, 0);
  buffer.writeUInt32BE(message.version, 1);

  return buffer;
}

function parseVersion(buffer) {
  return {
    messageType: 'V',
    version: buffer.readUInt32BE(1),
  };
}

function formatOrderAdded(message) {
  const buffer = Buffer.allocUnsafe(42);

  buffer.writeUInt8(0x41, 0);
  writeUInt64BE(buffer, message.timestamp, 1);
  writeUInt64BE(buffer, message.orderNumber, 9);
  writeString(buffer, message.side, 17, 1);
  writeString(buffer, message.instrument, 18, 8);
  writeUInt64BE(buffer, message.quantity, 26);
  writeUInt64BE(buffer, message.price, 34);

  return buffer;
}

function parseOrderAdded(buffer) {
  return {
    messageType: 'A',
    timestamp: readUInt64BE(buffer, 1),
    orderNumber: readUInt64BE(buffer, 9),
    side: readString(buffer, 17, 1),
    instrument: readString(buffer, 18, 8),
    quantity: readUInt64BE(buffer, 26),
    price: readUInt64BE(buffer, 34),
  };
}

function formatOrderExecuted(message) {
  const buffer = Buffer.allocUnsafe(29);

  buffer.writeUInt8(0x45, 0);
  writeUInt64BE(buffer, message.timestamp, 1);
  writeUInt64BE(buffer, message.orderNumber, 9);
  writeUInt64BE(buffer, message.quantity, 17);
  buffer.writeUInt32BE(message.matchNumber, 25);

  return buffer;
}

function parseOrderExecuted(buffer) {
  return {
    messageType: 'E',
    timestamp: readUInt64BE(buffer, 1),
    orderNumber: readUInt64BE(buffer, 9),
    quantity: readUInt64BE(buffer, 17),
    matchNumber: buffer.readUInt32BE(25),
  };
}

function formatOrderCanceled(message) {
  const buffer = Buffer.allocUnsafe(25);

  buffer.writeUInt8(0x58, 0);
  writeUInt64BE(buffer, message.timestamp, 1);
  writeUInt64BE(buffer, message.orderNumber, 9);
  writeUInt64BE(buffer, message.canceledQuantity, 17);

  return buffer;
}

function parseOrderCanceled(buffer) {
  return {
    messageType: 'X',
    timestamp: readUInt64BE(buffer, 1),
    orderNumber: readUInt64BE(buffer, 9),
    canceledQuantity: readUInt64BE(buffer, 17),
  };
}

function writeUInt64BE(buffer, value, offset) {
  buffer.writeUInt32BE(value / 0x100000000, offset);
  buffer.writeUInt32BE(value % 0x100000000, offset + 4);
}

function readUInt64BE(buffer, offset) {
  const high = buffer.readUInt32BE(offset);
  const low = buffer.readUInt32BE(offset + 4);

  return 0x100000000 * high + low;
}

function writeString(buffer, value, offset, length) {
  const count = buffer.write(value, offset, length, 'ascii');

  buffer.fill(0x20, offset + count, offset + length);
}

function readString(buffer, offset, length) {
  return buffer.toString('ascii', offset, offset + length);
}
